### PR TITLE
fabricstore: initial commit

### LINKED ISSUE #130: store: study the possibility of a fabricstore
Connected to a HyperLedger Fabric Blockchain.

### PR DESCRIPTION
**Pull request by piedup**
Resolves #130

### CODE REVIEW

**Comment by piedup** [comment]
I have tried to maximise test coverage using mocks. Coverage should increase once integration test for Hyperledger Fabric is included in sdk.

File: cmd/fabricstore/main.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+	a, err := fabricstore.New(&fabricstore.Config{
+		ChannelID:   *channelID,
+		ChaincodeID: *chaincodeID,
+		ConfigFile:  *configFile,
+		Version:     version,
+		Commit:      commit,
+	})
+	if err != nil {
+		fmt.Println("Could not start fabric client", err.Error())
```
Comment (such) [inline cmd/fabricstore/main.go]: I feel like in go, the idiom is to write:
```
if err != nil {
	log.Fatalf("Could not start fabric client: %v", err)
}
storehttp.RunWithFlags(a)
```

File: fabricstore/chaincode/pop/pop.go [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+		return s.GetValue(APIstub, args)
+	case "DeleteValue":
+		return s.DeleteValue(APIstub, args)
+	default:
+		return shim.Error("Invalid Smart Contract function name: " + function)
+	}
+}
+
+// SaveMap saves map into CouchDB using map document
+func (s *SmartContract) SaveMap(stub shim.ChaincodeStubInterface, segment *cs.Segment) error {
```
Comment (such) [inline fabricstore/chaincode/pop/pop.go]: does this function have to be public?

File: fabricstore/fabricstore.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+const (
+	// Name is the name set in the store's information.
+	Name = "fabric"
+
+	// Description is the description set in the store's information.
+	Description = "Stratumn Fabric Store"
+)
+
+// Config contains configuration optios for the store
```
Comment (such) [inline fabricstore/fabricstore.go]: typo: options

File: fabricstore/fabricstore.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+type Config struct {
+	// ChannelID used to send transactions
+	ChannelID string
+
+	// ChaincodeID used for transactions
+	ChaincodeID string
+
+	// ConfigFile path to netork configuration file (yaml)
```
Comment (such) [inline fabricstore/fabricstore.go]: typo: network

File: fabricstore/chaincode/pop/pop.go [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+	return shim.Success(nil)
+}
+
+// Invoke method is called as a result of an application request to run the Smart Contract "pop"
+func (s *SmartContract) Invoke(APIstub shim.ChaincodeStubInterface) sc.Response {
+	// Retrieve the requested Smart Contract function and arguments
+	function, args := APIstub.GetFunctionAndParameters()
+
+	switch function {
+	case "GetSegment":
```
Comment (such) [inline fabricstore/chaincode/pop/pop.go]: I think it'd be a bit cleaner to use const for this strings

File: fabricstore/integration/config-client.yaml [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+  cryptoconfig:
+    path: $GOPATH/src/github.com/stratumn/sdk/fabricstore/integration/crypto-config
+
+  tls:
+    enabled: false
+
+  credentialStore:
+    path: /tmp/.hfc-key-store
+    # path: $GOPATH/src/github.com/stratumn/sdk/fabricstore/integration/creds
```
Comment (such) [inline fabricstore/integration/config-client.yaml]: what should we keep?

File: fabricstore/integration/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    volumes:
+        - /var/run/:/host/var/run/
+        - ./../chaincode/:/opt/gopath/src/github.com/
+        - ./crypto-config:/opt/gopath/src/github.com/hyperledger/fabric/peer/crypto/
+    networks:
+        - basic
+    #depends_on:
+    #  - orderer.example.com
+    #  - peer0.org1.example.com
+    #  - couchdb
```
Comment (such) [inline fabricstore/integration/docker-compose.yml]: can we remove this?

File: fabricstore/integration/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,123 @@
+#
```
Comment (such) [inline fabricstore/integration/docker-compose.yml]: should we add our license header?

**Comment by piedup** [comment]
Thank you. I have integrated all your comments.

**Pull request merged**